api/shared/src/main/scala/dagr/api/models/TaskInfo.scala

```diff
@@ -0,0 +1,153 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+package dagr.api.models
+
+import dagr.api.DagrApi.TaskId
+import upickle.Js
+import upickle.default.{Reader, Writer}
+
+import scala.collection.mutable.ListBuffer
+import java.time.Instant
+
+/** The execution information for a task.  Used for extrenal read-only access to [[TaskInfo]]. Any execution system
+  * should extend [[TaskInfo]] instead class to store their specific metadata. */
+trait TaskInfoLike  {
+  def id         : Option[TaskId]
+  def attempts   : Int
+  def script     : Option[String]
+  def log        : Option[String]
+  def resources  : Option[ResourceSet]
+  def exitCode   : Option[Int]
+  def throwable  : Option[Throwable]
+
+  /** The current status of the task. */
+  def status     : TaskStatus
+
+  /** The instant the task reached a given status. */
+  def timePoints :  Traversable[TimePoint]
+
+  /** The instant the task reached the current status. */
+  def statusTime : Instant
+
+  /** The name of the task. */
+  def name: String
+
+  protected[dagr] def infoString: String = {
+    val resourceMessage = this.resources match {
+      case Some(r) => s"with ${r.cores} cores and ${r.memory} memory"
+      case None    => ""
+    }
+    s"'$name' : ${this.status} on attempt #${this.attempts} $resourceMessage"
+  }
+}
+
+case class TaskInfo
+(
+  name           : String,
+  id             : Option[TaskId],
+  attempts       : Int,
+  script         : Option[String],
+  log            : Option[String],
+  resources      : Option[ResourceSet],
+  exitCode       : Option[Int],
+  throwable      : Option[Throwable],
+  status         : TaskStatus,
+  timePoints     : Traversable[TimePoint],
+  statusTime     : Instant,
+  // Optional fields, if empty, no meaning should be inferred
+  scriptContents : Option[String] = None,
+  logContents    : Option[String] = None,
+  dependsOn      : Seq[TaskId]  = Seq.empty,
+  dependents     : Seq[TaskId] = Seq.empty,
+  parent         : Option[TaskId]  = None,
+  children       : Seq[TaskId]   = Seq.empty
+) extends TaskInfoLike {
+
+  def compare(that: TaskInfo): Int = {
+    (this.id, that.id) match {
+      case (Some(_), None)    => -1
+      case (None, Some(_))    => 1
+      case (None, None)       => this.status.ordinal - that.status.ordinal
+      case (Some(l), Some(r)) => (l - r).toInt
+    }
+  }
+}
+
+object TaskInfo {
+  import DagrApiDef._
+
+  implicit val query2Writer: Writer[TaskInfo] = Writer[TaskInfo] {
```

Comment:
You have to put these implicit `val`s in the companion object for pickling with `upickle`.